### PR TITLE
Unit testing framework for MarketAnalysis module

### PR DESCRIPTION
**Pull request by laxdog**
Start of the unit test work for https://github.com/Mikadily/poloniexlendingbot/issues/229

## Description
To use it, 

`pip install pytest hypothesis`

then

`pytest --verbose tests/`

## TESTING STAGE
No testing!

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] **I have read CONTRIBUTING.md**
- [x] **I fully understand [Github Flow.](https://guides.github.com/introduction/flow/)**
- [x] **My code adheres to the [code style of this project.](https://poloniexlendingbot.readthedocs.io/en/latest/contributing.html)**
- [N/A] **I have updated the documentation in /docs if I have changed the config, arguments, logic in how the bot works, or anything that understandably needs a documentation change.**
- [N/A] **I have updated the config file accordingly if my change requires a new configuration setting or changes an existing one.**
- [N/A] **I have tested the bot with no issues for 24 continuous hours. If issues were experienced, they have been patched and tested again.**

### CODE REVIEW

**Comment by laxdog** [comment]
Pretty much rewriting this at this point. I've moved the whole MarketAnalysis module to an actual class. Most of the tests are broken at the minute since I was working on the percentile stuff. 

I'll fix that class, I want to move the starting of the thread to a separate function and a few other bits and pieces. 

Though after that it'll need reviewed and possibly merged. I'd like to fix the other modules to work in the same way if everyone agrees. i.e. make them actual python classes and split the functions up a bit more to facilitate testing.

**Comment by rnevet** [comment]
Hi, just a small thing, aren't you making your changes based on old code?

**Comment by laxdog** [comment]
Shit. Maybe.

**Comment by rnevet** [comment]
Percentile issues were already fixed and you have a merge conflict.

**Comment by laxdog** [comment]
Yea, I made this branch ages ago and forgot to rebase. Bah! My own fault. OK, I'll fix it.

**Comment by rnevet** [comment]
Sorry for the late notice. ;)

**Comment by laxdog** [comment]
Right, here's an example where that method doesn't return the same as numpy:

```
> /Users/mrobinson/source/python/poloniexlendingbot/tests/test_MarketAnalysis.py(71)test_get_percentile()
-> assert(np_perc == math_perc)
(Pdb) math_perc
0.0
(Pdb) np_perc
1e-05
(Pdb) rates
[0.0, 0.00025, 0.00025]
(Pdb) lending_style
1
```

This is fixed in my commit.

**Comment by laxdog** [comment]
Is there a way to turn off the assert check for a certain folder @Evanito ? The whole tests folder is going to have these errors every time.

**Comment by laxdog** [comment]
I think this is done now. Definitely needs reviewed though.

File: modules/Lending.py [hunk 1]
```diff
@@ -171,9 +171,9 @@ def lend_all():
         lending_balances = Data.get_on_order_balances()
 
     # Fill the (maxToLend) balances on the botlog.json for display it on the web
-    for key in sorted(total_lended):
-        if len(lending_balances) == 0 or key not in lending_balances:
-            MaxToLend.amount_to_lend(total_lended[key], key, 0, 0)
+    for cur in sorted(total_lended):
```
Comment (rnevet) [inline modules/Lending.py]: @laxdog even though you are 100% right to change this name, I would avoid **next time** doing naming changes as part of such a big change.

**Comment by rnevet** [comment]
Looks very good, my only comment is that I needed to install pytz.
'pip install pytz'

File: modules/Lending.py (hunk 1 above)
Comment (laxdog) [inline modules/Lending.py]: Yea, sorry about that. I think I learned a lot of lessons during this PR with things not to do.

**Comment by laxdog** [comment]
I didn't realise about the pytz dependency. I'll write that up somewhere. 

Cheers for merging, I can get on with the other stuff now :)

**Pull request merged**